### DISCUSSION TITLE
Some confusion about channel.close()

### DISCUSSION THREAD

**Discussion by Yuan-zs** [rabbitmq/rabbitmq-server#10079] [Questions] [answered]
Hi all new here, I'm a bit confused and I hope to get your help. Encountered network issues when channel.close(), then may thorw IOException or TimeoutException and channel may not closed？I came up with an idea that, like this:
```
do {
  try{
     channel.close();
     //closed successfully，break the do-while
    break;
    } catch (IOException | TimeoutExcepiton e) {
       //channel may not closed, continue loop
       continue；
    } catch(ShutSingleException e) {
      //channel is closed，break the do-while
      break;
    }
 }while(true);  
```
I just want to sure channel is closed, otherwise it will no longer be useful, but it will continue to exist. I don't know if it's normal for me to do this, but I feel like it's just closing the channel specifically for loop. Is it a bit redundant? May I ask if there is a better approach?
Thanks in advance

**Comment by michaelklishin** [comment, marked as answer]
@Yuan-zs if a channel runs into a network delay, you can treat it as closed because when [its connection is closed by the heartbeat mechanism](https://rabbitmq.com/nettick.html), so will automatically be all the channels on it.

In fact, see an alternative to close called `abort` in the Java client ;)

**Reply by Yuan-zs** [reply]
Thank you for your answer，can I understand it as when `channel. close()`, and then throw `IOException`/`TimeoutException`. it can consider the connection to have been closed at this point? I looked the docs that the next time the connection is restored, the channel will also be restored. In that case, the channel will not be closed in the end？
Thanks very much

**Reply by michaelklishin** [reply]
A closed connection that throws an `IOException` almost always will be closed. The only exception is when [an alarm is in effect](https://rabbitmq.com/alarms.html) and the connection was publishing messages so it was blocked.

A [recovered connection](https://rabbitmq.com/api-guide.html#recovery) is a new connection on the wire. And the channels on it are different channels (even if their ID/numbers are identical).

**Reply by Yuan-zs** [reply]
Thanks for your providing the document. Throws an `IOException` almost always will be closed. But `factory.setAutomaticRecoveryEnabled(true);` can recover the connection and channel. Is my understanding correct?

Channel needs to be closed. But an IOException was thrown, indicating that both the connection and channel have been closed. But because `factory. setAutomaticRecoveryEnable (true)` the channel also will be restored, so there will still exist a channel causing resource leakage?(Although the channels are different channels (even if their ID/numbers are identical), there is a channel exist.)

**Reply by michaelklishin** [reply]
@Yuan-zs old channel objects won't have any references to them any more. I suggest that you read the [automatic connection recovery code](https://github.com/rabbitmq/rabbitmq-java-client/tree/main/src/main/java/com/rabbitmq/client/impl/recovery) instead of guessing what kind of "magic" may be happening there.

**Reply by Yuan-zs** [reply]
OK, thanks.
